---
title: "Basics of {rrrSnvs}"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Basics of rrrSnvs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r libraries}
library(Seurat)
library(rrrSnvs)
library(tidyverse)
```

## Using SNPs to define tumor clusters
```{r eval=FALSE}
two_sobj <- qs::qread("/home/gdrobertslab/mvc002/analyses/roberts/dev/testSnps/output/patient_met_1/two_sobj.qs")

# Do cell type annotation so I can use macrophages and monocytes as known normal cells
# The idea is that Osteosarcoma cells are unlikely to be mistakenly annotated as macrophages or monocytes
hpca <- celldex::HumanPrimaryCellAtlasData()
imm_cells <- celldex::MonacoImmuneData()
blueprint <- celldex::BlueprintEncodeData()

cell_assign <-
    SingleR::SingleR(as.SingleCellExperiment(two_sobj),
                     ref = list(hpca,
                                imm_cells,
                                blueprint),
                     labels = list(hpca$label.main,
                                   imm_cells$label.main,
                                   blueprint$label.main))

two_sobj$cell_type <-
    cell_assign$labels

two_sobj$cell_score <-
    cell_assign$scores %>%
    apply(MARGIN = 1, function(x) max(x, na.rm = TRUE))

c_b_t <-
    two_sobj@meta.data %>%
    select(used_clusters) %>%
    dplyr::rename(cell_group = used_clusters) %>%
    rownames_to_column("cell_barcode") %>%
    as_tibble() %>%
    mutate(bam_file = paste0("/home/gdrobertslab/lab/Counts/",
                             str_remove(cell_barcode, "_.*"),
                             "/outs/possorted_genome_bam.bam"),
           cell_barcode = str_remove(cell_barcode, ".+_"))

snp_tree <-
    get_snp_tree(cellid_bam_table = c_b_t,
                 ploidy = "GRCh37",
                 output_dir = "output",
                 output_base_name = "test_two_sobj",
                 ref_fasta = "/home/gdrobertslab/lab/GenRef/10x-human/fasta/genome.fa",
                 min_depth = c(5, 10, 20),
                 sbatch_base = "two_sobj",
                 min_snvs_per_cluster = 1000,
                 max_prop_missing_at_site = 0.9,
                 cleanup = FALSE)

# Pngs of trees and text output is in output/ directory

# Add snv groups to the Seurat object
two_sobj <-
    add_snv_group_to_sobj(
        two_sobj,
        snv_group_file = "output/test_two_sobj_5_dist.txt",
        new_column = "snv_group_5"
    ) %>%
    add_snv_group_to_sobj(
        "output/test_two_sobj_10_dist.txt",
        new_column = "snv_group_10"
    ) %>%
    add_snv_group_to_sobj(
        "output/test_two_sobj_20_dist.txt",
        new_column = "snv_group_20"
    )

DimPlot(
    two_sobj,
    group.by = c("snv_group_5", "snv_group_10", "snv_group_20", "cell_type"),
    label = TRUE,
    repel = TRUE,
    ncol = 2
) +
    NoLegend()

control_celltypes <- c("Monocytes", "Macrophages", "NK cells", "NK_cell")

# Figure out which clusters are more than 50% control celltypes
normal_clusters <-
    match_celltype_clusters(sobject = two_sobj,
                            normal_celltypes = control_celltypes,
                            cluster_col = "used_clusters",
                            celltype_col = "cell_type")

two_sobj <-
    label_tumor_cells(
        two_sobj,
        group_col_name = "used_clusters",
        snv_group_col = "snv_group_20",
        normal_clusters = normal_clusters,
        tumor_call_column = "snv_tumor_call_20"
    )

DimPlot(
    two_sobj,
    group.by = c("snv_group_20", "snv_tumor_call_20", "cell_type"),
    label = TRUE,
    repel = TRUE,
    ncol = 2
) +
    NoLegend()

```